visualizer.py

```diff
@@ -0,0 +1,311 @@
+#!/usr/bin/python3
+"""
+Author: Vertigo Designs, Ryan Summers
+
+Description: Bokeh application for serving Stabilizer stream visuals.
+"""
+import argparse
+import asyncio
+import enum
+import logging
+import socket
+from typing import List, Mapping
+
+import bokeh.plotting
+import bokeh.layouts
+import bokeh.document
+import bokeh.palettes
+import bokeh.io
+import bokeh.models
+import bokeh.server.server
+import numpy as np
+
+from miniconf import Miniconf
+from stabilizer.stream import StabilizerStream
+
+# The sample rate of stabilizer.
+SAMPLE_RATE_HZ = 100e6 / 128
+
+
+def _get_ip(broker):
+    """ Get the IP of the local device.
+
+    Args:
+        broker: The broker IP of the test. Used to select an interface to get the IP of.
+
+    Returns:
+        The IP as an array of integers.
+    """
+    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    try:
+        sock.connect((broker, 1883))
+        address = sock.getsockname()[0]
+    finally:
+        sock.close()
+
+    return list(map(int, address.split('.')))
+
+
+class TriggerState(enum.Enum):
+    """ The state of the scope trigger. """
+    IDLE = enum.auto()
+    ARMED = enum.auto()
+    TRIGGERED = enum.auto()
+    STOPPED = enum.auto()
+
+
+class StreamReceiver:
+    """ Handles asynchronously receiving Stabilizer stream frames. """
+
+    def __init__(self, port):
+        """ Initialize the receiver on the provided port. """
+        self.traces = dict()
+        self.trace_index = 0
+        self.times = []
+        self._current_format = None
+
+        self.trigger = TriggerState.IDLE
```

Comment:
This is [MVVM style](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93viewmodel). The `StreamReceiver` is the model, `StreamVisualizer` is the modelview, and the actual JS frontend is the view. I think this would be clearer if I did some class renaming.